src/php/Compiler/Domain/Parser/ExpressionParser/AtomParser.php

```diff
@@ -44,50 +48,33 @@ public function parse(Token $token): AbstractAtomNode
             return $this->parseKeyword($token);
         }
 
-        if (preg_match('/^([+-])?0[bB][01]+(_[01]+)*$/', $word, $matches)) {
-            // binary numbers
-            $sign = (isset($matches[1]) && $matches[1] === '-') ? -1 : 1;
-            return new NumberNode(
-                $word,
-                $token->getStartLocation(),
-                $token->getEndLocation(),
-                $sign * bindec(str_replace('_', '', $word)),
-            );
+        if (preg_match(self::REGEX_BINARY_NUMBER, $word, $matches)) {
+            return $this->parseBinaryNumber($matches, $word, $token);
         }
 
-        if (preg_match('/^([+-])?0[xX][0-9a-fA-F]+(_[0-9a-fA-F]+)*$/', $word, $matches)) {
-            // hexdecimal numbers
-            $sign = (isset($matches[1]) && $matches[1] === '-') ? -1 : 1;
-            return new NumberNode(
-                $word,
-                $token->getStartLocation(),
-                $token->getEndLocation(),
-                $sign * hexdec(str_replace('_', '', $word)),
-            );
+        if (preg_match(self::REGEX_HEXADECIMAL_NUMBER, $word, $matches)) {
+            return $this->parseHexadecimalNumber($matches, $word, $token);
         }
 
-        if (preg_match('/^([+-])?0[0-7]+(_[0-7]+)*$/', $word, $matches)) {
-            // octal numbers
-            $sign = (isset($matches[1]) && $matches[1] === '-') ? -1 : 1;
-            return new NumberNode(
-                $word,
-                $token->getStartLocation(),
-                $token->getEndLocation(),
-                $sign * octdec(str_replace('_', '', $word)),
-            );
+        if (preg_match(self::REGEX_OCTAL_NUMBER, $word, $matches)) {
+            return $this->parseOctalNumber($matches, $word, $token);
         }
 
         if (is_numeric($word)) {
             return new NumberNode($word, $token->getStartLocation(), $token->getEndLocation(), $word + 0);
         }
 
+        if (preg_match(self::REGEX_DECIMAL_NUMBER, $word, $matches)) {
+            return $this->parseDecimalNumber($matches, $word, $token);
+        }
+
```

Comment:
Not sure if I understood properly.
What is the difference between the two "numeric" methods (method from L63 & L67)?
Are not both of them returning a `new NumberNode()`? Maybe both conditions could be simplify into one?

I have done some tests with `is_numeric()` PHP method and it's interesting:
```php
$tests = array(
    123,     # is_numeric is TRUE
    1_23,    # is_numeric is TRUE
    ' 123 ', # is_numeric is TRUE
    -1_23,   # is_numeric is TRUE
    '1_23 ', # is_numeric is FALSE
);

foreach ($tests as $element) {
    if (is_numeric($element)) {
        echo var_export($element, true) . ' is numeric', PHP_EOL;
    } else {
        echo var_export($element, true) . ' is NOT numeric', PHP_EOL;
    }
}
```

Reply:
You answered yourself: `is_numeric('1_23') # is_numeric is FALSE`, 
but it's a valid decimal number with underscore.

Reply:
What I mean is if it maybe makes sense to do something like this:
```php
public function parse(Token $token): AbstractAtomNode
{
    // ...

    if ($this->isNumeric($word)) {
        return $this->parseDecimalNumber($matches, $word, $token);
    }

    // ...
}

private function isNumeric(string $word): bool
{
    return is_numeric($word)
        || preg_match(self::REGEX_DECIMAL_NUMBER, $word, $matches);
}
```

Reply:
Oh, I didn't get it before. Good idea! Feel free to commit and push this yourself in this branch :)